### PR TITLE
Further improve error message for E0081

### LINKED ISSUE #97533: Improve diagnostics for duplicate enum discriminants
This is split out from the discussion in #97319.

First, the existing error for duplicate enum discriminants could be improved. The best summary of what we have now can be seen in the test added for #97319, which fixes part of it:

```rust
enum Enum {
    P = 3,
    X = 3,
    Y = 5
}

#[repr(u8)]
enum EnumOverflowRepr {
    P = 257,
    X = 513,
}

#[repr(i8)]
enum NegDisEnum {
    First = -1,
    Second = -2,
    Last,
}
```

Which outputs:

```
error[E0081]: discriminant value `3` assigned more than once
  --> $DIR/E0081.rs:1:1
   |
LL | / enum Enum {
LL | |
LL | |     P = 3,
   | |         - first assignment of `3`
LL | |
LL | |     X = 3,
   | |         - second assignment of `3`
LL | |
LL | |     Y = 5
LL | | }
   | |_^

error[E0081]: discriminant value `1` assigned more than once
  --> $DIR/E0081.rs:11:1
   |
LL | / enum EnumOverflowRepr {
LL | |
LL | |     P = 257,
   | |         --- first assignment of `1` (overflowed from `257`)
LL | |
LL | |     X = 513,
   | |         --- second assignment of `1` (overflowed from `513`)
LL | |
LL | | }
   | |_^

error[E0081]: discriminant value `-1` assigned more than once
  --> $DIR/E0081.rs:20:1
   |
LL | / enum NegDisEnum {
LL | |
LL | |     First = -1,
   | |             -- first assignment of `-1`
LL | |
LL | |     Second = -2,
   | |     ----------- assigned discriminant for `Last` was incremented from this discriminant
LL | |
LL | |     Last,
   | |     ---- second assignment of `-1`
LL | |
LL | | }
   | |_^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0081`.
```

I suggested that the last version of the error should suggest explicitly stating that the discriminant for `Last` is computed as `(Second = -2) + (1 variant later) = -1 = First`, or somehow better wording that. Essentially, make it clear to the user that things are incremented positively, regardless of the order of existing discriminants.

Secondly, @Bryysen suggested that the code that does this (`check_enum`) could definitely use some additional refactoring as well in [this comment](https://github.com/rust-lang/rust/issues/97319#issuecomment-1140519399).

### PR DESCRIPTION
**Pull request by Bryysen**
Closes #97533

### CODE REVIEW

File: compiler/rustc_typeck/src/check/check.rs [hunk 1]
```diff
@@ -1494,76 +1493,96 @@ fn check_enum<'tcx>(tcx: TyCtxt<'tcx>, vs: &'tcx [hir::Variant<'tcx>], def_id: L
...
+                {
+                    let ve_ident = var.ident;
+                    let sp = if n > 1 { "variants" } else { "variant" };
+                    let n = n + 1;
+
+                    err.span_label(
+                        *span,
+                        format!("discriminant for `{ve_ident}` incremented from this startpoint (`{ident}` + {n} {sp} later => `{ve_ident}` = {dis})"),
+                    );
+                }
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/check.rs:1536]: Should the case where there is only implicit discriminants be handled here?

File: compiler/rustc_typeck/src/check/check.rs [hunk 2]
```diff
@@ -1494,76 +1493,96 @@ fn check_enum<'tcx>(tcx: TyCtxt<'tcx>, vs: &'tcx [hir::Variant<'tcx>], def_id: L
...
+
+                (vs[idx].span, format!("`{dis}`"))
+            }
+        };
+
+        err.span_label(span, format!("{display_discr} assigned here"));
+    };
+
+    let mut i = 0;
+    while i < discrs.len() {
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/check.rs:1546]: The structure of the algorithm, with a mutable `discrs` is not really obvious. Could you document how this works?

File: compiler/rustc_typeck/src/check/check.rs [hunk 3]
```diff
@@ -1494,76 +1493,96 @@ fn check_enum<'tcx>(tcx: TyCtxt<'tcx>, vs: &'tcx [hir::Variant<'tcx>], def_id: L
...
+                    && let rustc_ast::LitKind::Int(lit_value, _int_kind) = &lit.node
+                    && *lit_value != dis.val
+                {
+                    (tcx.hir().span(expr.hir_id), format!("`{dis}` (overflowed from `{lit_value}`)"))
+                } else {
+                    (tcx.hir().span(expr.hir_id), format!("`{dis}`"))
+                }
+            }
+            None => {
+                if let Some((n, hir::Variant { span, ident, .. })) =
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/check.rs:1525]: Could you add a comment please?

File: compiler/rustc_typeck/src/check/check.rs [hunk 4]
```diff
@@ -1494,76 +1493,96 @@ fn check_enum<'tcx>(tcx: TyCtxt<'tcx>, vs: &'tcx [hir::Variant<'tcx>], def_id: L
...
 
     check_representable(tcx, sp, def_id);
     check_transparent(tcx, sp, def);
 }
 
-/// In the case that a discriminant is both a duplicate and an overflowing literal,
-/// we insert both the assigned discriminant and the literal it overflowed from into the formatted
-/// output. Otherwise we format the discriminant normally.
-fn format_discriminant_overflow<'tcx>(
+fn detect_discriminant_duplicate<'tcx>(
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/check.rs:1502]: Could you add a comment please?

File: compiler/rustc_typeck/src/check/check.rs [hunk 5]
```diff
@@ -1494,76 +1493,96 @@ fn check_enum<'tcx>(tcx: TyCtxt<'tcx>, vs: &'tcx [hir::Variant<'tcx>], def_id: L
...
-        if let hir::ExprKind::Lit(lit) = &body.kind
-            && let rustc_ast::LitKind::Int(lit_value, _int_kind) = &lit.node
-            && dis.val != *lit_value
-        {
-                    return format!("`{dis}` (overflowed from `{lit_value}`)");
+    mut discrs: Vec<(VariantIdx, Discr<'tcx>)>,
+    vs: &'tcx [hir::Variant<'tcx>],
+    self_span: Span,
+) {
+    let report = |var: &hir::Variant<'_>,
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/check.rs:1508]: This closure already captures `vs` and has `idx` passed. Can it compute `var = vs[idx]` on its own?

File: src/test/ui/error-codes/E0081.stderr [hunk 6]
```diff
@@ -29,14 +29,35 @@ LL | enum NegDisEnum {
    | ^^^^^^^^^^^^^^^
 LL |
 LL |     First = -1,
-   |             -- first assignment of `-1`
+   |             -- `-1` assigned here
 LL |
 LL |     Second = -2,
-   |     ----------- assigned discriminant for `Last` was incremented from this discriminant
+   |     ----------- discriminant for `Last` incremented from this startpoint (`Second` + 1 variant later => `Last` = -1)
```
Comment (cjgillot) [inline src/test/ui/error-codes/E0081.stderr:35]: Is there a way to have this explanation appear *after* the note on `Last`?

**Review by cjgillot — COMMENTED** [review]
Thanks @Bryysen!
Could you add a little PR message to explain how you are improving the message?
I left a few comments in code too.

**Comment by Bryysen** [comment]
Sorry, was too hasty when I made the PR :smile:.

So in the old algorithm we would do a lazy evaluation of the discriminant values, and store them in a vec as we go. Once we evaluate a new discriminant, we would compare its value to the previously computed discriminants and if it matched any of them, we error with a duplicate. The issue with this approach is that we can't know ahead of time all the duplicates for a given discriminant `N` due to the lazy evaluation, and so we would generate a seperate error for every duplicate match. This [comment](https://github.com/rust-lang/rust/issues/97533#issuecomment-1140532094) on the linked issue illustrates this.

My solution was to just eagerly evaluate all the discriminants, and then do a similar comparison between all discriminants as we did before. When we hit a duplicate this time, we add to the error message the span + value and [shuffle that discriminant out](https://github.com/Bryysen/rust/blob/4ee2fe308bc68a0d93a86a09a1fd2506cc41e9b2/compiler/rustc_typeck/src/check/check.rs#L1573) of the `vec` to avoid erroring on it multiple times, then continue the cycle to round up any other duplicates for that discriminant value.

Complexity wise, this should be no worse than the old algorithm (unless I've misjugded something). Hope that clears some things up. I'll take a second to go through the review comments.

>Could you add a little PR message to explain how you are improving the message?

The biggest improvement is reducing the verbosity for multiple duplicates of the same discriminant value. Another small improvement is that we point out where (and how) an implicit duplicate is incremented from, which can hopefully help clear confusion for new users:

```
#[repr(i8)]                                                  
enum MyEnum {                                                      
    V0 = 0,                                                  
    V1 = -2,                                                                                   
    V2,                                
    V3,               
    V4 = 0,
} 
```
```
error[E0081]: discriminant value `0` assigned more than once
  --> tst.rs:5:1
   |
5  | enum MyEnum {
   | ^^^^^^^^^^^
6  |     V0 = 0,
   |          - `0` assigned here
7  |     V1 = -2,
   |     ------- discriminant for `V3` incremented from this startpoint (`V1` + 2 variant later => `V3` = 0)
8  |     V2,
9  |     V3,
   |     -- `0` assigned here
10 |     V4 = 0,
   |          - `0` assigned here

error: aborting due to previous error
```
However, i'm unsure about the ```discriminant for `V3` incremented from this startpoint (`V1` + 2 variant later => `V3` = 0)``` part, maybe there is a better way to word this?

File: src/test/ui/error-codes/E0081.stderr (hunk 6 above)
Comment (Bryysen) [reply src/test/ui/error-codes/E0081.stderr:35]: We should be able to do something like that, although I'm not that well versed in the diagnostic part of rustc. One thing we could do is split the explanation into its own span:
```
error[E0081]: discriminant value `-1` assigned more than once
 --> tst.rs:5:1
  |
5 | enum NegDisEnum {
  | ^^^^^^^^^^^^^^^
6 |     First = -1,
  |             -- `-1` assigned here
7 |     Second = -2,
8 |     Last,
  |     ---- `-1` assigned here
  |
note: discriminant for `Last` incremented from this startpoint (`Second` + 1 variant later => `Last` = -1)
 --> tst.rs:7:5
  |
7 |     Second = -2,
  |     ^^^^^^^^^^^

error: aborting due to previous error
```
The drawback is that this adds slightly to diagnostic clutter.

File: compiler/rustc_typeck/src/check/check.rs (hunk 5 above)
Comment (Bryysen) [reply compiler/rustc_typeck/src/check/check.rs:1508]: You are right! We don't need to pass it as an argument, fixed

File: compiler/rustc_typeck/src/check/check.rs (hunk 4 above)
Comment (Bryysen) [reply compiler/rustc_typeck/src/check/check.rs:1502]: I've documunted the more obscure code in the function, let me know if i'm missing something

File: compiler/rustc_typeck/src/check/check.rs (hunk 3 above)
Comment (Bryysen) [reply compiler/rustc_typeck/src/check/check.rs:1525]: Added some comments, will push changes soon

File: compiler/rustc_typeck/src/check/check.rs (hunk 2 above)
Comment (Bryysen) [reply compiler/rustc_typeck/src/check/check.rs:1546]: Added comments to (hopefully) clarify whats going on, will push the changes soon

File: compiler/rustc_typeck/src/check/check.rs (hunk 1 above)
Comment (Bryysen) [reply compiler/rustc_typeck/src/check/check.rs:1536]: Not sure I understand, could you clarify?

**Comment by cjgillot** [comment]
Thanks @Bryysen!
@bors r+

**Comment by Bryysen** [comment]
Would you like me to squash the commits first? And thanks for taking the time to review!

**Pull request merged**